### PR TITLE
Jumpdest constructor inj

### LINKED ISSUE #2607: jump opcode analysis
### Description
Every time an executable account (contract) is run, a brand new Code object is instantiated from the bytecode in the account. Upon encountering the first JUMP instruction, all valid jump locations are computed and stored in a bitmap on the Code object. Unfortunately, that Code object is not retained for future use, so this analysis must be repeated on the first JUMP instruction of every contract, every time it is run.

### Acceptance Criteria

1. Reuse of jump destinations on all subsequent contract executions. 
2. Runtime speed improvements proportional to the number of jump dests in the contract. Pathological EVMtool test case [attached](https://gist.github.com/jflo/6f46fd44fd706c3f26deaacadf5b1a45), should speed up considerably.
3. Cache entries must be evicted on Account deletion.
4. Cache entries must be evicted on contract self-destruct.
5. Entries should expire out of cache in LRU order when total cache weight is hit.

### PR DESCRIPTION
**Pull request by jflo**
## PR description
Alternate implementation of [this PR](https://github.com/hyperledger/besu/pull/2809), which does not require a singleton for the jumpdest cache.
## Fixed Issue(s)

fixes #2607 

## Changelog

- [X] I thought about the changelog and included a [changelog update if required](https://wiki.hyperledger.org/display/BESU/Changelog).

### CODE REVIEW

File: besu/src/main/java/org/hyperledger/besu/cli/options/unstable/ContractCacheOptions.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+import org.hyperledger.besu.evm.internal.JumpDestCacheConfiguration;
+
+import java.util.Arrays;
+import java.util.List;
+
+import picocli.CommandLine;
+
+public class ContractCacheOptions implements CLIOptions<JumpDestCacheConfiguration> {
+
+  public static final String CONTRACT_CACHE_WEIGHT = "--Xcontract-code-cache-weight-kb";
```
Comment (jflo) [inline besu/src/main/java/org/hyperledger/besu/cli/options/unstable/ContractCacheOptions.java:28]: The internals of this cache have gotten more specific. Should we rename the option for accuracy, or would that be detail the user doesn't need?

File: evm/src/main/java/org/hyperledger/besu/evm/EVM.java [hunk 2]
```diff
@@ -150,4 +159,33 @@ public Operation operationAtOffset(final Code code, final int offset) {
...
+   * @param code The code within which we are looking for the destination.
+   * @return Whether or not this location is a valid jump destination.
+   */
+  public boolean isValidJumpDestination(final UInt256 destination, final Code code) {
+    if (!destination.fitsInt()) return false;
+
+    final int jumpDestination = destination.intValue();
+    if (jumpDestination >= code.getSize()) return false;
+    long[] validJumpDestinations = code.getValidJumpDestinations();
+    if (validJumpDestinations == null || validJumpDestinations.length == 0) {
```
Comment (shemnon) [inline evm/src/main/java/org/hyperledger/besu/evm/EVM.java:176]: If we are not putting the bits back on the `Code` object then this branch will always be visited.  We should either set it on the Code object or move the bit calculation code here and only look to the cache for the bits.

**Review by shemnon — APPROVED** [review]
Looks better.  Moving the code bit calculation logic is a minor issue that doesn't (for me) block merging.

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/bonsai/BonsaiLayeredWorldState.java [hunk 3]
```diff
@@ -42,7 +42,6 @@
   protected final long height;
   protected final TrieLogLayer trieLog;
   private final Hash worldStateRootHash;
-
```
Comment (macfarla) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/bonsai/BonsaiLayeredWorldState.java:45]: spurious change

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/ClassicProtocolSpecs.java [hunk 4]
```diff
@@ -209,17 +235,21 @@ public static ProtocolSpecBuilder phoenixDefinition(
...
             configStackSizeLimit,
             enableRevertReason,
             ecip1017EraRounds,
-            quorumCompatibilityMode)
+            quorumCompatibilityMode,
+            jumpdestCacheConfiguration)
         .gasCalculator(IstanbulGasCalculator::new)
         .evmBuilder(
-            gasCalculator -> MainnetEVMs.istanbul(gasCalculator, chainId.orElse(BigInteger.ZERO)))
+            (gasCalculator, jdCacheConf) ->
```
Comment (macfarla) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/ClassicProtocolSpecs.java:250]: is jd shortening it too much

File: ethereum/core/src/test/java/org/hyperledger/besu/ethereum/vm/EVMTest.java [hunk 5]
```diff
@@ -42,20 +44,22 @@
...
   @Before
   public void setup() {
-    evm = new EVM(operationRegistry, gasCalculator);
+    evm = new EVM(operationRegistry, gasCalculator, JumpDestCacheConfiguration.DEFAULT_CONFIG);
   }
 
   @Test
   public void assertThatEndOfScriptNotExplicitlySetInCodeReturnsAVirtualOperation() {
-    final Code code = new Code(Bytes.fromHexString("0x60203560003555606035604035556000"));
+    Bytes noEnd = Bytes.fromHexString("0x60203560003555606035604035556000");
```
Comment (macfarla) [inline ethereum/core/src/test/java/org/hyperledger/besu/ethereum/vm/EVMTest.java:52]: final

File: ethereum/core/src/test/java/org/hyperledger/besu/ethereum/vm/EVMTest.java [hunk 6]
```diff
@@ -42,20 +44,22 @@
...
+    final Code code = new Code(noEnd, Hash.hash(noEnd));
     final Operation operation = evm.operationAtOffset(code, code.getSize());
     assertThat(operation).isNotNull();
     assertThat(operation.isVirtualOperation()).isTrue();
   }
 
   @Test
   public void assertThatEndOfScriptExplicitlySetInCodeDoesNotReturnAVirtualOperation() {
-    final Code code = new Code(Bytes.fromHexString("0x6020356000355560603560403555600000"));
+    Bytes ends = Bytes.fromHexString("0x6020356000355560603560403555600000");
```
Comment (macfarla) [inline ethereum/core/src/test/java/org/hyperledger/besu/ethereum/vm/EVMTest.java:61]: final

File: ethereum/core/src/test/java/org/hyperledger/besu/ethereum/vm/operations/JumpOperationTest.java [hunk 7]
```diff
@@ -78,16 +83,17 @@ public void init() {
...
     registry.put(new JumpOperation(gasCalculator));
     registry.put(new JumpDestOperation(gasCalculator));
-    evm = new EVM(registry, gasCalculator);
+    evm = new EVM(registry, gasCalculator, JumpDestCacheConfiguration.DEFAULT_CONFIG);
   }
 
   @Test
   public void shouldJumpWhenLocationIsJumpDest() {
     final JumpOperation operation = new JumpOperation(gasCalculator);
+    Bytes jumpBytes = Bytes.fromHexString("0x6003565b00");
```
Comment (macfarla) [inline ethereum/core/src/test/java/org/hyperledger/besu/ethereum/vm/operations/JumpOperationTest.java:92]: final

File: ethereum/core/src/test/java/org/hyperledger/besu/ethereum/vm/operations/JumpOperationTest.java [hunk 8]
```diff
@@ -112,41 +119,73 @@ public void shouldJumpWhenLocationIsJumpDestAndAtEndOfCode() {
...
-            .code(new Code(Bytes.fromHexString("0x6801000000000000000c565b00")))
+            .code(new Code(jumpBytes, Hash.hash(jumpBytes)))
             .build();
     frameDestinationGreaterThanCodeSize.setPC(CURRENT_PC);
 
-    final OperationResult result = operation.execute(frameDestinationGreaterThanCodeSize, null);
+    final OperationResult result = operation.execute(frameDestinationGreaterThanCodeSize, evm);
     assertThat(result.getHaltReason()).contains(ExceptionalHaltReason.INVALID_JUMP_DESTINATION);
-
+    Bytes badJump = Bytes.fromHexString("0x60045600");
```
Comment (macfarla) [inline ethereum/core/src/test/java/org/hyperledger/besu/ethereum/vm/operations/JumpOperationTest.java:132]: final

File: ethereum/core/src/test/java/org/hyperledger/besu/ethereum/vm/operations/JumpOperationTest.java [hunk 9]
```diff
@@ -112,41 +119,73 @@ public void shouldJumpWhenLocationIsJumpDestAndAtEndOfCode() {
   @Test
   public void shouldHaltWithInvalidJumDestinationWhenLocationIsOutsideOfCodeRange() {
     final JumpOperation operation = new JumpOperation(gasCalculator);
+    Bytes jumpBytes = Bytes.fromHexString("0x6801000000000000000c565b00");
```
Comment (macfarla) [inline ethereum/core/src/test/java/org/hyperledger/besu/ethereum/vm/operations/JumpOperationTest.java:122]: final

File: ethereum/referencetests/src/main/java/org/hyperledger/besu/ethereum/referencetests/ReferenceTestCode.java [hunk 10]
```diff
@@ -32,6 +33,6 @@
    */
   @JsonCreator
   public ReferenceTestCode(final String bytes) {
-    super(Bytes.fromHexString(bytes));
+    super(Bytes.fromHexString(bytes), Hash.EMPTY);
```
Comment (macfarla) [inline ethereum/referencetests/src/main/java/org/hyperledger/besu/ethereum/referencetests/ReferenceTestCode.java:36]: do we not need the hash(bytes) here

**Review by macfarla — APPROVED** [review]
LGTM - a few comments

File: besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java [hunk 11]
```diff
@@ -1735,6 +1736,7 @@ public BesuControllerBuilder getControllerBuilder() {
                 .orElse(GasLimitCalculator.constant()))
         .requiredBlocks(requiredBlocks)
         .reorgLoggingThreshold(reorgLoggingThreshold)
+        .jumpDestCacheConfiguration(ContractCacheOptions.create().toDomainObject())
```
Comment (antonydenyer) [inline besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java:1739]: use `unstableContractCacheOptions`

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/MainnetProtocolSchedule.java [hunk 12]
```diff
@@ -60,11 +64,29 @@ public static ProtocolSchedule fromConfig(
...
    * @return A configured mainnet protocol schedule
    */
   public static ProtocolSchedule fromConfig(
-      final GenesisConfigOptions config, final boolean isRevertReasonEnabled) {
-    return fromConfig(config, PrivacyParameters.DEFAULT, isRevertReasonEnabled);
+      final GenesisConfigOptions config,
+      final boolean isRevertReasonEnabled,
+      final JumpDestCacheConfiguration jumpdestCacheConfiguration) {
+    return fromConfig(
+        config, PrivacyParameters.DEFAULT, isRevertReasonEnabled, jumpdestCacheConfiguration);
```
Comment (antonydenyer) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/MainnetProtocolSchedule.java:75]: /jumpdestCacheConfiguration/jumpDestCacheConfiguration/ for consistency?

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/MainnetProtocolSpecs.java [hunk 13]
```diff
@@ -162,6 +164,7 @@ public static ProtocolSpecBuilder frontierDefinition(
         .blockImporterBuilder(MainnetBlockImporter::new)
         .blockHeaderFunctions(new MainnetBlockHeaderFunctions())
         .miningBeneficiaryCalculator(BlockHeader::getCoinbase)
+        .jumpDestCacheConfig(jumpDestCacheConfiguration)
```
Comment (antonydenyer) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/MainnetProtocolSpecs.java:167]: /Config/Configuration/  

I know both are used in the codebase, but would be good if we could start being consistent at least within the same feature. p.s don't mind which!

File: besu/src/main/java/org/hyperledger/besu/cli/options/unstable/ContractCacheOptions.java (hunk 1 above)
Comment (antonydenyer) [reply besu/src/main/java/org/hyperledger/besu/cli/options/unstable/ContractCacheOptions.java:28]: Would it be useful to scope the wider configuration as something broader ie `EvmConfiguration`?  And inside that have something specific with an evm prefix?

**Review by antonydenyer — CHANGES_REQUESTED** [review]
Good to see that singleton got rid of :smiley: 
Couple of changes but nothing major.

File: besu/src/main/java/org/hyperledger/besu/cli/options/unstable/ContractCacheOptions.java (hunk 1 above)
Comment (jflo) [reply besu/src/main/java/org/hyperledger/besu/cli/options/unstable/ContractCacheOptions.java:28]: Good ideas all. I'm renaming both config and options classes. New cli flag will be `--Xevm-jumpdest-cache-weight-kb`

File: CHANGELOG.md [hunk 14]
```diff
@@ -5,6 +5,7 @@
 - The EVM has been factored out into a standalone module, suitable for inclusion as a library. [#2790](https://github.com/hyperledger/besu/pull/2790)
 - Low level performance improvements changes to cut worst-case EVM performance in half. [#2796](https://github.com/hyperledger/besu/pull/2796)
 - Migrate `ExceptionalHaltReason` from an enum to an interface to allow downstream users of the EVM to add new exceptional halt reasons. [#2810](https://github.com/hyperledger/besu/pull/2810)
+- reduces need for JUMPDEST analysis via caching [#2607](https://github.com/hyperledger/besu/issues/2607)
```
Comment (matkt) [inline CHANGELOG.md:8]: I used to put the links of PR and not issues here

File: besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java [hunk 15]
```diff
@@ -1587,7 +1589,6 @@ private void configure() throws Exception {
                     : SyncMode.FULL);
 
     ethNetworkConfig = updateNetworkConfig(getNetwork());
-
```
Comment (matkt) [inline besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java:1590]: I think this change is not useful for this PR

File: evm/src/main/java/org/hyperledger/besu/evm/EVM.java (hunk 2 above)
Comment (jflo) [reply evm/src/main/java/org/hyperledger/besu/evm/EVM.java:176]: The bits go back into the Code object [here](https://github.com/hyperledger/besu/blob/909df0cba9776f4713e3b74464cbb6094fe788b1/evm/src/main/java/org/hyperledger/besu/evm/Code.java#L115).

Comment (jflo) [reply evm/src/main/java/org/hyperledger/besu/evm/EVM.java:176]: I'm going to put some thought into this. 🤔 The Jump operations ask for this map from the EVM itself, not the Code.

**Review by matkt — APPROVED** [review]
LGTM

File: besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java [hunk 16]
```diff
@@ -1292,6 +1293,7 @@ private void handleUnstableOptions() {
             .put("Native Library", unstableNativeLibraryOptions)
             .put("Data Storage Options", unstableDataStorageOptions)
             .put("Launcher", unstableLauncherOptions)
+            .put("Jump Destination Cache", unstableEvmOptions)
```
Comment (antonydenyer) [inline besu/src/main/java/org/hyperledger/besu/cli/BesuCommand.java:1296]: EVM Configurations Options

File: besu/src/main/java/org/hyperledger/besu/controller/BesuControllerBuilder.java [hunk 17]
```diff
@@ -241,7 +248,7 @@ public BesuController build() {
     checkNotNull(nodeKey, "Missing node key");
     checkNotNull(storageProvider, "Must supply a storage provider");
     checkNotNull(gasLimitCalculator, "Missing gas limit calculator");
-
+    checkNotNull(evmConfiguration, "Missing jumpdest cache config");
```
Comment (antonydenyer) [inline besu/src/main/java/org/hyperledger/besu/controller/BesuControllerBuilder.java:251]: Missing EVM Configuration

File: besu/src/test/java/org/hyperledger/besu/ForkIdsTest.java [hunk 18]
```diff
@@ -156,7 +157,8 @@ public void testForkId() {
     final GenesisConfigFile genesisConfigFile =
         GenesisConfigFile.fromConfig(EthNetworkConfig.jsonConfig(chainName));
     final GenesisConfigOptions configOptions = genesisConfigFile.getConfigOptions();
-    final ProtocolSchedule schedule = MainnetProtocolSchedule.fromConfig(configOptions);
+    final ProtocolSchedule schedule =
+        MainnetProtocolSchedule.fromConfig(configOptions, EvmConfiguration.DEFAULT_CONFIG);
```
Comment (antonydenyer) [inline besu/src/test/java/org/hyperledger/besu/ForkIdsTest.java:161]: `EvmConfiguration.DEFAULT` to be consistent with others?

File: ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/ClassicProtocolSpecs.java [hunk 19]
```diff
@@ -209,17 +229,21 @@ public static ProtocolSpecBuilder phoenixDefinition(
...
             configStackSizeLimit,
             enableRevertReason,
             ecip1017EraRounds,
-            quorumCompatibilityMode)
+            quorumCompatibilityMode,
+            evmConfiguration)
         .gasCalculator(IstanbulGasCalculator::new)
         .evmBuilder(
-            gasCalculator -> MainnetEVMs.istanbul(gasCalculator, chainId.orElse(BigInteger.ZERO)))
+            (gasCalculator, jumpDestCacheConf) ->
```
Comment (antonydenyer) [inline ethereum/core/src/main/java/org/hyperledger/besu/ethereum/mainnet/ClassicProtocolSpecs.java:244]: Can you check the params are correct here?

**Review by antonydenyer — APPROVED** [review]
Couple of nits but all good otherwise

File: evm/src/main/java/org/hyperledger/besu/evm/internal/EvmConfiguration.java [hunk 20]
```diff
@@ -0,0 +1,33 @@
...
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.hyperledger.besu.evm.internal;
+
+public class EvmConfiguration {
```
Comment (matkt) [inline evm/src/main/java/org/hyperledger/besu/evm/internal/EvmConfiguration.java:18]: have you decided to not use the Immutable library for this object ?

Comment (jflo) [reply evm/src/main/java/org/hyperledger/besu/evm/internal/EvmConfiguration.java:18]: Not specifically. I'll add that in a later pr.

**Pull request merged**